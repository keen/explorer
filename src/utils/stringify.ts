import { TIMEZONES } from '@keen.io/query-creator';

export const stringify = (queryParams) =>
  Object.keys(queryParams)
    .map((k) => {
      let queryParamValue = queryParams[k];
      if (!queryParamValue) return null;
      if (Array.isArray(queryParamValue) && !queryParamValue.length)
        return null;

      if (k === 'timezone') {
        const timezoneOption = TIMEZONES.find(
          (item) => item.name === queryParamValue
        ) || {
          label: 'UTC',
          value: 0,
        };
        queryParamValue = timezoneOption.value;
      }

      if (Array.isArray(queryParamValue)) {
        queryParamValue = queryParamValue.map((value) => {
          if (typeof value === 'object' && value !== null) {
            const underscoredObject = {};
            Object.keys(value).forEach((objkey) => {
              const underscoredK = objkey
                .replace(/(?:^|\.?)([A-Z])/g, (x, y) => `_${y.toLowerCase()}`)
                .replace(/^_/, '');
              underscoredObject[underscoredK] = value[objkey];
            });
            return underscoredObject;
          }
          return value;
        });
      }

      queryParamValue = JSON.stringify(queryParamValue);

      const underscoredK = k
        .replace(/(?:^|\.?)([A-Z])/g, (x, y) => `_${y.toLowerCase()}`)
        .replace(/^_/, '');
      return `${encodeURIComponent(underscoredK)}=${encodeURIComponent(
        queryParamValue
      )}`;
    })
    .filter((item) => !!item)
    .join('&')
    .replace(/%22true%22/gi, 'true')
    .replace(/%22false%22/gi, 'false');
