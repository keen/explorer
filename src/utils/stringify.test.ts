/* eslint-disable @typescript-eslint/camelcase */
import { stringify } from './stringify';

test('stringify query params', () => {
  const queryParams = {
    filters: [],
    timezone: 'UTC',
    timeframe: 'this_14_weeks',
    zero_fill: null,
    order_by: [
      { direction: 'ASC', property_name: 'result' },
      { direction: 'ASC', property_name: 'platform' },
    ],
    interval: null,
    group_by: ['referrer', 'platform'],
    limit: null,
    event_collection: 'purchases',
  };

  const queryParamsWithFilters = {
    steps: [
      {
        event_collection: 'purchases',
        actor_property: 'item.price',
        timeframe: 'this_14_days',
        filters: [],
        inverted: false,
        optional: false,
        with_actors: false,
      },
      {
        event_collection: 'signups',
        actor_property: 'user.age',
        timeframe: 'this_14_days',
        filters: [
          {
            operator: 'ne',
            property_type: 'String',
            property_value: 'male',
            property_name: 'user.gender',
          },
        ],
        inverted: false,
        optional: false,
        with_actors: false,
      },
    ],
  };

  expect(stringify(queryParams)).toMatchSnapshot();
  expect(stringify(queryParamsWithFilters)).toMatchSnapshot();
});
