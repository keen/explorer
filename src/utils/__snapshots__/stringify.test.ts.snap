// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringify query params 1`] = `"timezone=0&timeframe=%22this_14_weeks%22&order_by=%5B%7B%22direction%22%3A%22ASC%22%2C%22property_name%22%3A%22result%22%7D%2C%7B%22direction%22%3A%22ASC%22%2C%22property_name%22%3A%22platform%22%7D%5D&group_by=%5B%22referrer%22%2C%22platform%22%5D&event_collection=%22purchases%22"`;

exports[`stringify query params 2`] = `"steps=%5B%7B%22event_collection%22%3A%22purchases%22%2C%22actor_property%22%3A%22item.price%22%2C%22timeframe%22%3A%22this_14_days%22%2C%22filters%22%3A%5B%5D%2C%22inverted%22%3Afalse%2C%22optional%22%3Afalse%2C%22with_actors%22%3Afalse%7D%2C%7B%22event_collection%22%3A%22signups%22%2C%22actor_property%22%3A%22user.age%22%2C%22timeframe%22%3A%22this_14_days%22%2C%22filters%22%3A%5B%7B%22operator%22%3A%22ne%22%2C%22property_type%22%3A%22String%22%2C%22property_value%22%3A%22male%22%2C%22property_name%22%3A%22user.gender%22%7D%5D%2C%22inverted%22%3Afalse%2C%22optional%22%3Afalse%2C%22with_actors%22%3Afalse%7D%5D"`;
